File: Paragony/Services/ReceiptService.cs
using Microsoft.EntityFrameworkCore;
using Paragony.Abstract;
using Paragony.Data;
using Paragony.Models;

namespace Paragony.Services;

public class ReceiptService(AppDbContext context, IOpenAIService openAiService) : IReceiptService
{
    public async Task<Receipt> AddReceiptFromImage(string base64Image)
    {
        // Analyze receipt image using OpenAI
        var receipt = await openAiService.AnalyzeReceiptImage(base64Image);

        // Save image to disk
        var imageName = $"{Guid.NewGuid()}.jpg";
        var imagePath = Path.Combine("Uploads", imageName);
        Directory.CreateDirectory("Uploads");
        await File.WriteAllBytesAsync(imagePath, Convert.FromBase64String(base64Image));

        // Store the image file name in the receipt
        receipt.ImagePath = imageName;

        var categories = await context.Categories.ToListAsync();

        foreach (var item in receipt.Items)
        {
            var category = categories.FirstOrDefault(c => c.Name == item.CategoryName);

            if (category == null)
            {
                category = new Category
                {
                    Name = item.CategoryName,
                };
                await context.Categories.AddAsync(category);
                await context.SaveChangesAsync();
            }

            item.CategoryId = category.Id;
        }

        // Save to database
        context.Receipts.Add(receipt);
        await context.SaveChangesAsync();

        return receipt;
    }

    public async Task<List<Receipt>> GetAllReceipts(DateOnly? datefrom, DateOnly? dateTo)
    {
        var query = context.Receipts
            .Include(r => r.Items)
            .ThenInclude(i => i.Category)
            .AsQueryable();

        if (datefrom.HasValue) query = query.Where(r => r.PurchaseDate >= datefrom.Value);

        if (dateTo.HasValue) query = query.Where(r => r.PurchaseDate <= dateTo.Value);

        return await query
            .OrderByDescending(r => r.PurchaseDate)
            .ToListAsync();
    }

    public async Task<Receipt> GetReceiptById(Guid id)
    {
        return await context.Receipts
            .Include(x => x.Items)
            .ThenInclude(i => i.Category)
            .FirstOrDefaultAsync(r => r.Id == id) ?? throw new KeyNotFoundException("Receipt not found");
    }

    public async Task<decimal> GetSpendingByCategory(string category, DateOnly? startDate, DateOnly? endDate)
    {
        var query = context.ReceiptItems
            .Where(item => item.Category.Name == category);

        if (startDate.HasValue)
            query = query.Where(item => item.Receipt.PurchaseDate >= startDate.Value);

        if (endDate.HasValue)
            query = query.Where(item => item.Receipt.PurchaseDate <= endDate.Value);

        return await query.SumAsync(item => item.Price);
    }

    public async Task<decimal> GetSpendingByDate(DateOnly date)
    {
        return await context.Receipts
            .Where(r => r.PurchaseDate == date)
            .SumAsync(r => r.TotalAmount);
    }

    public async Task<List<Receipt>> SearchReceipts(string query)
    {
        return await context.Receipts
            .Include(r => r.Items)
            .ThenInclude(i => i.Category)
            .Where(r => r.StoreName.Contains(query) ||
                        r.Items.Any(i => i.Name.Contains(query) || i.Category.Name.Contains(query)))
            .ToListAsync();
    }

    public async Task<List<ReceiptItem>> SearchSpendingsByStoreName(string query)
    {
        return await context.ReceiptItems
            .Include(item => item.Receipt)
            .Where(item => item.Receipt.StoreName.ToLower().Contains(query.ToLower()))
            .ToListAsync();
    }

    public async Task<Receipt?> GetLastReceiptContainingKeyword(string query)
    {
        return await context.Receipts
            .Include(r => r.Items)
            .ThenInclude(i => i.Category)
            .Where(r => r.StoreName.Contains(query) ||
                        r.Items.Any(i => i.Name.Contains(query) || i.Category.Name.Contains(query)))
            .OrderByDescending(r => r.PurchaseDate)
            .FirstOrDefaultAsync();
    }
}

File: agents/receipt_service.py
from sqlalchemy.orm import Session
from models import Receipt, ReceiptItem, Category
from datetime import date
from sqlalchemy import func
import logging

class ReceiptService:
    def __init__(self, db_session: Session):
        self.db_session = db_session

    def get_all_receipts(self, date_from: date = None, date_to: date = None):
        query = self.db_session.query(Receipt).join(Receipt.items).join(ReceiptItem.category)

        if date_from:
            query = query.filter(Receipt.purchase_date >= date_from)
        if date_to:
            query = query.filter(Receipt.purchase_date <= date_to)

        return query.order_by(Receipt.purchase_date.desc()).all()

    def get_receipt_by_id(self, receipt_id):
        receipt = self.db_session.query(Receipt).filter(Receipt.id == receipt_id).first()
        if not receipt:
            raise ValueError("Receipt not found")
        return receipt

    def get_spending_by_category(self, category_name, start_date: date = None, end_date: date = None):
        logging.info(f"Querying database for category: {category_name}, start_date: {start_date}, end_date: {end_date}")

        # Ensure the join logic reflects the actual relationships
        query = self.db_session.query(func.sum(ReceiptItem.price)).join(Receipt).join(Category).filter(Category.name == category_name)

        if start_date:
            query = query.filter(Receipt.purchase_date >= start_date)
        if end_date:
            query = query.filter(Receipt.purchase_date <= end_date)

        result = query.scalar() or 0
        logging.info(f"Database response for category '{category_name}': {result}")
        return result

    def get_spending_by_date(self, specific_date: date):
        return self.db_session.query(func.sum(Receipt.total_amount)).filter(Receipt.purchase_date == specific_date).scalar() or 0

    def search_receipts(self, search_query):
        return self.db_session.query(Receipt).join(Receipt.items).join(ReceiptItem.category).filter(
            (Receipt.store_name.contains(search_query)) |
            (ReceiptItem.name.contains(search_query)) |
            (Category.name.contains(search_query))
        ).all()

    def get_last_receipt_containing_keyword(self, keyword):
        return self.db_session.query(Receipt).join(Receipt.items).join(ReceiptItem.category).filter(
            (Receipt.store_name.contains(keyword)) |
            (ReceiptItem.name.contains(keyword)) |
            (Category.name.contains(keyword))
        ).order_by(Receipt.purchase_date.desc()).first()

# Example usage:
# from database import SessionLocal
# db_session = SessionLocal()
# receipt_service = ReceiptService(db_session)
# receipts = receipt_service.get_all_receipts()
